skrifa/src/outline/mod.rs

```diff
@@ -554,16 +572,120 @@ mod tests {
                     &mut path,
                 )
                 .unwrap();
-            if path.elements != expected_outline.path {
-                panic!(
-                    "mismatch in glyph path for id {} (size: {}, coords: {:?}): path: {:?} expected_path: {:?}",
+            assert_eq!(path.elements, expected_outline.path, "mismatch in glyph path for id {} (size: {}, coords: {:?}): path: {:?} expected_path: {:?}",
                     expected_outline.glyph_id,
                     expected_outline.size,
                     expected_outline.coords,
                     &path.elements,
                     &expected_outline.path
                 );
+        }
+    }
+
+    #[derive(Copy, Clone, Debug, PartialEq)]
+    enum GlyphPoint {
+        On { x: f32, y: f32 },
+        Off { x: f32, y: f32 },
+    }
+
+    #[derive(Debug)]
+    struct PointCaptor {
+        points: Vec<GlyphPoint>,
+    }
+
+    impl PointCaptor {
+        fn new() -> Self {
+            Self { points: Vec::new() }
+        }
+
+        fn into_points(mut self) -> Vec<GlyphPoint> {
+            // if we start and end with the same on-curve we can drop the end
```

Comment:
that's ok (the fonttools SegmentToPointPen also does the same) -- however there is an important point to bear in mind when converting from segment to points as you drop the last point overlapping the first. If you want to be able to round-trip back from points to segment, which is an equally important thing to be wishing to do, when you draw a list of points as segments (moveTo, lineTo, quadTo, etc.), you need to take care of explicitly emitting the last lineTo command which would be implied when the contour is closed ('Z' or closePath). The PointToSegmentPen has an option that when True it unconditionally emits these; when False (default) it only emits them for closed contours if the last and first point do overlap (imagine a four points that in some location look like a triangle with duplicate points at the end and then can morph into a rectangle); it does this to disambiguate this occurrence from the implied closing lineTo. If it did not do that, the duplicate points would be lost when converted back to segments.

Reply:
in write-fonts::tables::glyf::simple, where we have the opposite problem of converting a BezPath (sequence of curve _segments_) into a stream of points, we have this special handling for the duplicate last point in closed contours, see the test here:

https://github.com/googlefonts/fontations/blob/0cf2a550ceb7a6c2830767da9311a66fd1c11ccd/write-fonts/src/tables/glyf/simple.rs#L1285-L1331

it's only natural that the respective read_fonts::tables::glyf::to_path (which converts from point stream to segments or path drawing commands) does the equivalent of this but on the reverse.

Reply:
Good point, I will rig a test and fix.

Reply:
Looks to me like pens are always advised of a closing line, even if it's a nop. Added the test `always_get_closing_lines` to prove it and ensure it remains so.

Added `triangle_and_square_retain_compatibility` to prove our triangle and square are both captured as 4 points. PointPen's handling of close now has a more limited drop-last with rationale. Ptal.